File: sqlColumn.java
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import py4j.GatewayServer;
import com.alibaba.druid.sql.SQLUtils;
import com.alibaba.druid.sql.ast.SQLStatement;
import com.alibaba.druid.sql.visitor.SchemaStatVisitor;


public class sqlColumn {

    private final String dbType;

    public sqlColumn(String sqltype) throws Exception {
        Class classType = Class.forName("com.alibaba.druid.util.JdbcConstants");
        String dbType = classType.getField(sqltype).get(null).toString();   //数据库类型
        this.dbType = dbType;
    }

    public Map<String, List> getColumn(String sql) throws Exception {
        try {
            List<SQLStatement> stmtList = SQLUtils.parseStatements(sql, dbType);
            SQLStatement stmt = stmtList.get(0);
            
            SchemaStatVisitor statVisitor = SQLUtils.createSchemaStatVisitor(dbType);
            stmt.accept(statVisitor);
            
            Map<String, List> dict = new HashMap<String, List>();
            dict.put("tables", statVisitor.getTables().keySet().stream().map(x -> x.toString()).collect(Collectors.toList()));
            dict.put("columns", statVisitor.getColumns().stream().map(x -> x.toString()).collect(Collectors.toList()));
            dict.put("conditions", statVisitor.getConditions().stream().map(x -> x.toString()).collect(Collectors.toList()));
            return dict;
        } 
        
        catch (Exception e) {
            System.out.println("SOMETHING WRONG");
            throw e;
        }
    }

    public static void main(String[] args) throws Exception {
        sqlColumn obj = new sqlColumn("MYSQL");
        GatewayServer server = new GatewayServer(obj, 25335);
        server.start();
        /*String testsql = "update dwd_info_invest2 c,\n" +
                "\t\t(select \n" +
                "        (@i := case when @pre_part_no=key_account then @i + 1 else 1 end) row_number,     \n" +
                "\t\t\t\tdwd_info_invest.key_primary,\n" +
                "        (@pre_part_no:=key_account)        \n" +
                "    from dwd_info_invest,\n" +
                "    (SELECT @i := 0, @pre_part_no:='') AS a\n" +
                "\t  where status_invest <> 3\n" +
                "    group by key_account, dt_invest, key_primary  \n" +
                "    order by key_account) b\n" +
                "set c.rank_allinvest = b.row_number\n" +
                "where c.key_primary = b.key_primary;";
        System.out.println(obj.getColumn(testsql));
        */
    }

}

File: sqlColumn.scala
import py4j.GatewayServer
import com.alibaba.druid.sql.SQLUtils
import scala.collection.JavaConverters._


class sqlGetColumn(sql_type: String) {

  private val type_class = Class.forName("com.alibaba.druid.util.JdbcConstants")
  val dbType = type_class.getField(sql_type).get(null).toString()

  def get_column(sql: String): java.util.Map[String,java.util.List[String]] = {
    val stmtList = SQLUtils.parseStatements(sql, dbType)
    val stmt = stmtList.get(0)
    val statVisitor = SQLUtils.createSchemaStatVisitor(dbType)
    stmt.accept(statVisitor)
    Map(
      "tables" -> statVisitor.getTables().keySet().asScala.map(x => x.toString).toList.asJava,
      "columns" -> statVisitor.getColumns().asScala.map(x => x.toString).toList.asJava
    ).asJava
  }

}

object sqlColumn {
  def main(args: Array[String]): Unit = {
    val obj = new sqlGetColumn("MYSQL")
    val server = new GatewayServer(obj, 25336)
    server.start()
    //val testsql = "UPDATE dwd_info_invest2 c,(SELECT ( @i := CASE WHEN @pre_part_no = key_account THEN @i + 1 ELSE 1 END ) row_number, dwd_info_invest.key_primary, ( @pre_part_no := key_account ) FROM dwd_info_invest, ( SELECT @i := 0, @pre_part_no := '' ) AS a WHERE status_invest <> 3 AND type_sourcetable IN ( 1, 2, 4 ) GROUP BY key_account,dt_invest,key_primary ORDER BY key_account ) b SET c.rank_partinvest = b.row_number WHERE c.key_primary = b.key_primary;"
    //val res1 = obj.get_column(testsql)
    //println(res1)
  }
}
